### PR TITLE
fix QUOTED_STRING and BLOCK_STRING tokenizing problem

### PR DESCRIPTION
**Pull request by tompng**
QUOTED_STRING and BLOCK_STRING does not seem to be tokenized correctly.

```ruby
GraphQL::VERSION
# => "1.8.7"

# QUOTED_STRING
GraphQL::Language::Lexer.run_lexer '"\\\\", "\\"", "\\"\\\\"'
# => [(STRING "\\\", " [1:1]), (UNKNOWN_CHAR "\\" [1:7]), (STRING "" [1:8]), (STRING "\"\\" [1:11])]

# BLOCK_STRING
GraphQL::Language::Lexer.run_lexer %|""""a""","""b"""|
# => [(STRING "\"a\"\"\", \"\"\"b" [1:1])] (should be [(STRING "a"), (STRING "b")]

GraphQL::Language::Lexer.run_lexer %|"""\\n"""|
# => [(STRING "\n" [1:1])] (should be "\\" + "n". should not be unescaped)

GraphQL::Language::Lexer.run_lexer %|""""""|
GraphQL::Language::Lexer.run_lexer %|"""\n\n\n"""|
# => NoMethodError: undefined method `empty?' for nil:NilClass
```

This pullrequest:
- fixes definition of QUOTED_STRING and BLOCK_STRING
- fixes NoMethodError when tokenizing empty BLOCK_STRING

# Quoted String
quoted string consists of escaped characters(`\` + any) and normal characters(not `\"`)
```
BACKSLASH_ESCAPED_CHAR = '\\' /./;
STRING_CHAR =   (BACKSLASH_ESCAPED_CHAR | [^\\"]);
```

# Block String
any string that does not end with `"` or ends with `"""` can be explicitly splitted into these 3 groups.
BLOCK_QUOTE `"""`
BLOCK_STRING_ESCAPE_FRAGMENT `\+` `"\+` `""\+`
BLOCK_STRING_FRAGMENT `[^\"]+` `"[^\"]+` `""[^\"]+`

and the restrictions are:
BLOCK_QUOTE should appear after BLOCK_STRING_ESCAPE_FRAGMENT
should not end with BLOCK_STRING_ESCAPE_FRAGMENT
```
SHORT_BLOCK_QUOTES = ('' | '"' | '""');
BLOCK_STRING_FRAGMENT = SHORT_BLOCK_QUOTES [^\\"]+;
BLOCK_STRING_ESCAPE_FRAGMENT = SHORT_BLOCK_QUOTES '\\'+;
BLOCK_STRING_CHAR = (BLOCK_STRING_ESCAPE_FRAGMENT+ '"""' | BLOCK_STRING_ESCAPE_FRAGMENT* BLOCK_STRING_FRAGMENT);
```

### CODE REVIEW

**Comment by rmosolgo** [comment]
Wow, thanks for tackling this! Impressive work. It makes sense to me but let me take a closer look soon.

**Comment by tompng** [comment]
I found a performance problem in my pull requst.
```ruby
quote_string = '"' + '\\"' * 1000 + '"'
block_string = '"""' + '\\""\\"""' * 1000 + '"""'
# valid QUOTED_STRING and BLOCK_QUOTE (fast)
Benchmark.measure{GraphQL::Language::Lexer.tokenize(quote_string)}.real
# => 0.0026719998568296432
Benchmark.measure{GraphQL::Language::Lexer.tokenize(block_string)}.real
# => 0.006445999955758452

# invalid QUOTED_STRING and BLOCK_QUOTE (very slow)
Benchmark.measure{GraphQL::Language::Lexer.tokenize(quote_string[0...-1])}.real
# => 0.7374779998790473
Benchmark.measure{GraphQL::Language::Lexer.tokenize(quote_string[0...-1])}.real
# => 2.590942000038922
```

I think something like this is happening inside.
```
string "\"\"\"\"
will be check if it is:
QUOTE_STRING → fail
UNKNOWN_CHAR(") UNKNOWN_CHAR(\) QUOTE_STRING → fail
UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) QUOTE_STRING → fail
UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) QUOTE_STRING → fail
UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) UNKNOWN_CHAR(") UNKNOWN_CHAR(\) → tokenize success(takes long time to reach here)
```

I have no idea how to fix this by changing the definition of QUOTED_STRING and BLOCK_STRING.
(I think the problem is not in the definition of STRING_CHAR and BLOCK_STRING_CHAR)
However, raising error immediately and stop tokenizing if it find UNKNOWN_CHAR might solve this.
```ruby
UNKNOWN_CHAR => { raise TokenizeError, meta }; # dont emit :UNKNOWN_CHAR. just stop tokenizing.
```

**Comment by tompng** [comment]
```ruby
s = '"""' + '\\"""""' * 1000 + '\\""'
GraphQL::Language::Lexer.tokenize s
```

will be tokenized to
`""`, `"\""`, `""`, `"\""`, ..., `""`, `"\""`, `""`
(there is no UNKNOWN_CHAR, but it takes very long time)

Maybe adding UNMATCHED_BLOCK_QUOTE and UNMATCHED_QUOTED_STRING will solve this performance problem.
(If it starts with '"""', it will always be BLOCK_QUOTE or UNMATCHED_BLOCK_QUOTE)
```
QUOTED_STRING = QUOTE STRING_CHAR* QUOTE;
BLOCK_STRING = BLOCK_QUOTE BLOCK_STRING_CHAR* BLOCK_QUOTE;
UNMATCHED_QUOTED_STRING = QUOTE STRING_CHAR*;
UNMATCHED_BLOCK_STRING = BLOCK_QUOTE BLOCK_STRING_CHAR* SHORT_BLOCK_QUOTES;

UNMATCHED_BLOCK_STRING => { emit(:UNMATCHED_BLOCK_STRING, ts, te, meta) };
UNMATCHED_QUOTED_STRING => { emit(:UNMATCHED_QUOTED_STRING, ts, te, meta) };
```
(I will update this pull request later)

File: lib/graphql/language/lexer.rl [hunk 1]
```diff
@@ -100,6 +104,9 @@
 
     BLANK   => { meta[:col] += te - ts };
 
+
+    UNMATCHED_BLOCK_STRING => { emit(:UNMATCHED_BLOCK_STRING, ts, te, meta) };
+    UNMATCHED_QUOTED_STRING => { emit(:UNMATCHED_QUOTED_STRING, ts, te, meta) };
```
Comment (rmosolgo) [inline lib/graphql/language/lexer.rl]: What happens when these tokens are passed to the parser? I guess we get some kind of error, but is it a helpful one? Could we add a test to describe this case?

**Comment by tompng** [comment]
Performance issue that I commented seems to be solved.
```ruby
quote_string = '"' + '\\"' * 1000 + '"'
block_string = '"""' + '\\""\\"""' * 1000 + '"""'
# invalid QUOTED_STRING and BLOCK_QUOTE
Benchmark.measure{GraphQL::Language::Lexer.tokenize(quote_string[0...-1])}.real
# => 0.0018109995871782303 (was 0.7374779998790473 in my previous commit)
Benchmark.measure{GraphQL::Language::Lexer.tokenize(quote_string[0...-1])}.real
# => 0.001803000457584858 (was 2.590942000038922 in my previous commit)
```

```ruby
s = '"""' + '\\"""""' * 1_000_000 + '\\""'; s.size #=> 6000006
Benchmark.measure{GraphQL::Language::Lexer.tokenize s}.real #=> 4.885995999909937
# It's fast enough (compared to parsing a string "aaaaa..." in version 1.8.11)
Benchmark.measure{GraphQL::Language::Lexer.tokenize 'a'*6000006}.real #=> 4.973197000101209
```

**Comment by rmosolgo** [comment]
Hey, I'm really sorry that this has sat unresolved for so long. After you identified the performance issue, I forgot to come back and see that you fixed it 😖  Sorry! 

Are you still interested in making this change? The lexer has changed a bit to handle some _different_ edge cases, so there are conflicts. I'm happy to merge it if you want to resolve the conflicts, or work on a new fix from `master`.

**Comment by rmosolgo** [comment]
Please feel free to open a new PR with these changes against master. I'm so sorry I let this fall behind :confounded:

**Pull request closed without merging**